kotlin-inject-compiler/ksp/src/main/kotlin/me/tatarka/inject/compiler/ksp/ProcessTargetComponentAccessor.kt

```diff
@@ -0,0 +1,60 @@
+package me.tatarka.inject.compiler.ksp
+
+import com.google.devtools.ksp.processing.CodeGenerator
+import com.google.devtools.ksp.symbol.KSFunctionDeclaration
+import com.squareup.kotlinpoet.ksp.writeTo
+import me.tatarka.inject.compiler.COMPONENT
+import me.tatarka.inject.compiler.TargetComponentAccessorGenerator
+import me.tatarka.kotlin.ast.AstClass
+import me.tatarka.kotlin.ast.AstFunction
+import me.tatarka.kotlin.ast.KSAstProvider
+
+internal fun processTargetComponentAccessor(
+    element: KSFunctionDeclaration,
+    provider: KSAstProvider,
+    codeGenerator: CodeGenerator,
+    targetComponentAccessorGenerator: TargetComponentAccessorGenerator,
+    skipValidation: Boolean = false,
+): Boolean = with(provider) {
+    val astFunction = element.toAstFunction()
+    val returnType = astFunction.returnType
+    val returnTypeClass by lazy { returnType.resolvedType().toAstClass() }
+
+    // the generated actual function will be annotated with TargetComponentAccessor
+    // KSP will process them as well so we need to ignore them
+    if (astFunction.isActual && element.findExpects().firstOrNull() != null) return true
+    if (!astFunction.validateIsExpect(provider)) return true
+    if (!skipValidation && returnType.isError) return false
+    if (!astFunction.validateReturnType(returnTypeClass, provider)) return true
+
+    process(astFunction, returnTypeClass, codeGenerator, targetComponentAccessorGenerator)
+    true
+}
+
+private fun AstFunction.validateIsExpect(provider: KSAstProvider) = isExpect.also { isValid ->
+    if (!isValid) {
+        provider.error("$name should be an expect fun", this)
+    }
```

Comment:
I figured in this case it can't become an expect fun in later rounds anyways, so just report the issue right away. But I can move it to the generator.